File: elixir/hamming/lib/hamming.ex
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> Hamming.hamming_distance(~c"AAGTCATA", ~c"TAGCGATC")
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: {:ok, non_neg_integer} | {:error, String.t()}
  def hamming_distance(strand1, strand2) when length(strand1) != length(strand2) do
    {:error, "strands must be of equal length"}
  end

  def hamming_distance(strand1, strand2) do
    { :ok, Enum.zip_reduce(strand1, strand2, [],
                          (fn x, y, acc -> [x != y | acc] end))
    |> Enum.count(fn x -> x == true end) }
  end
end

File: erlang/hamming/src/hamming.erl
-module(hamming).

-export([distance/2]).


distance(S1, S2) when length(S1) =/= length(S2) ->
  {error,badarg};
distance(Strand1, Strand2) ->
  lists:sum(
    lists:zipwith(
      fun(S1, S2) ->
          case (S1 =:= S2) of
            true -> 0;
            false -> 1
          end
      end,
      Strand1,
      Strand2)).
